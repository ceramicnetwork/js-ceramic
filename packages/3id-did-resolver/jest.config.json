{
  "moduleFileExtensions": ["js", "json", "ts"],
  "testEnvironment": "node",
  "resolver": "jest-resolver-enhanced",
  "extensionsToTreatAsEsm": [".ts"],
  "globals": {
    "ts-jest": {
      "useESM": true
    }
  },
  "moduleNameMapper": {
    "^(\\.{1,2}/.*)\\.js$": "$1"
  },
  "testRegex": "(/__tests__/.*(\\.|/)(test|spec))\\.tsx?$"
}
