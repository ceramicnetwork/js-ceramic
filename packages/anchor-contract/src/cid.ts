import type { EventFragment } from '@ethersproject/abi'
import type { Log, TransactionResponse } from '@ethersproject/providers'
import { create as createMultihash } from 'multiformats/hashes/digest'
import { CID } from 'multiformats/cid'
import { fromString } from 'uint8arrays'

import { contractInterface } from './interface.js'

const SHA256_CODE = 0x12
const DAG_CBOR_CODE = 0x71
const V1_PROOF_TYPE = 'f(bytes32)'

export function createCidFromHexValue(value: string): CID {
  const multihash = createMultihash(SHA256_CODE, fromString(value.slice(2), 'base16'))
  return CID.create(1, DAG_CBOR_CODE, multihash)
}

export function getCidFromAnchorEventLog(log: Log): CID {
  const decodedArgs = contractInterface.decodeEventLog(
    contractInterface.events['DidAnchor'] as EventFragment,
    log.data
  )
  return createCidFromHexValue(decodedArgs[1])
}

const getCidFromV0Transaction = (txResponse: TransactionResponse): CID => {
  const withoutPrefix = txResponse.data.replace(/^(0x0?)/, '')
  return CID.decode(fromString(withoutPrefix.slice(1), 'base16'))
}

const getCidFromV1Transaction = (txResponse: TransactionResponse): CID => {
  const decodedArgs = contractInterface.decodeFunctionData('anchorDagCbor', txResponse.data)
  return createCidFromHexValue(decodedArgs[0])
}

/**
 * Parses the transaction data to recover the CID.
 * @param txType transaction type of the anchor proof. Currently support `raw` and `f(bytes32)`
 * @param txResponse the retrieved transaction from the ethereum blockchain
 * @returns
 */
export const getCidFromTransaction = (txType: string, txResponse: TransactionResponse): CID => {
  if (txType === V1_PROOF_TYPE) {
    return getCidFromV1Transaction(txResponse)
  } else {
    return getCidFromV0Transaction(txResponse)
  }
}
