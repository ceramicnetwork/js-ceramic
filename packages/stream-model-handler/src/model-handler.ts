import { Model } from '@ceramicnetwork/stream-model'
import {
  AnchorStatus,
  CommitData,
  CommitType,
  Context,
  SignatureStatus,
  SignatureUtils,
  StreamConstructor,
  StreamHandler,
  StreamState,
  StreamUtils,
} from '@ceramicnetwork/common'
import { StreamID } from '@ceramicnetwork/streamid'
import { SchemaValidation } from './schema-utils.js'
import { ViewsValidation } from './views-utils.js'
import { applyAnchorCommit } from '@ceramicnetwork/stream-handler-common'

// Keys of the 'ModelDefinition' type.  Unfortunately typescript doesn't provide a way to access
// these programmatically.
const ALLOWED_CONTENT_KEYS = new Set([
  'name',
  'description',
  'schema',
  'accountRelation',
  'relations',
  'views',
])

/**
 * Helper function for asserting that the content of a Model Stream only contains the expected fields
 */
const assertNoExtraKeys = function (content: Record<string, any>) {
  for (const key of Object.keys(content)) {
    if (!ALLOWED_CONTENT_KEYS.has(key)) {
      throw new Error(`Unexpected key '${key}' found in content for Model Stream`)
    }
  }
}

/**
 * Model stream handler implementation
 */
export class ModelHandler implements StreamHandler<Model> {
  private readonly _schemaValidator: SchemaValidation
  private readonly _viewsValidator: ViewsValidation

  get type(): number {
    return Model.STREAM_TYPE_ID
  }

  get name(): string {
    return Model.STREAM_TYPE_NAME
  }

  get stream_constructor(): StreamConstructor<Model> {
    return Model
  }

  constructor() {
    this._schemaValidator = new SchemaValidation()
    this._viewsValidator = new ViewsValidation()
  }

  /**
   * Applies commit (genesis|signed|anchor)
   * @param commitData - Commit (with JWS envelope or anchor proof, if available and extracted before application)
   * @param context - Ceramic context
   * @param state - Document state
   */
  async applyCommit(
    commitData: CommitData,
    context: Context,
    state?: StreamState
  ): Promise<StreamState> {
    if (process.env.CERAMIC_ENABLE_EXPERIMENTAL_COMPOSE_DB != 'true') {
      context.loggerProvider
        .getDiagnosticsLogger()
        .err(
          'Indexing is an experimental feature and is not yet supported in production. To enable for testing purposes only, set the CERAMIC_ENABLE_EXPERIMENTAL_COMPOSE_DB environment variable to `true`'
        )
      throw new Error('Indexing is not enabled')
    }

    if (state == null) {
      // apply genesis
      return this._applyGenesis(commitData, context)
    }

    if (StreamUtils.isAnchorCommitData(commitData)) {
      return this._applyAnchor(commitData, state)
    }

    throw new Error('Cannot update a finalized Model')
  }

  /**
   * Applies genesis commit
   * @param commitData - Genesis commit
   * @param context - Ceramic context
   * @private
   */
  async _applyGenesis(commitData: CommitData, context: Context): Promise<StreamState> {
    const payload = commitData.commit
    const isSigned = StreamUtils.isSignedCommitData(commitData)
    if (!isSigned) {
      throw Error('Model genesis commit must be signed')
    }

    if (!(payload.header.controllers && payload.header.controllers.length === 1)) {
      throw new Error('Exactly one controller must be specified')
    }

    const streamId = await StreamID.fromGenesis('model', commitData.commit)
    const { controllers, model } = payload.header
    const controller = controllers[0]
    const modelStreamID = StreamID.fromBytes(model)

    await SignatureUtils.verifyCommitSignature(
      commitData,
      context.did,
      controller,
      modelStreamID,
      streamId
    )

    assertNoExtraKeys(payload.data)
    Model.assertComplete(payload.data)

    const modelStreamId = StreamID.fromBytes(payload.header.model)
    if (!modelStreamId.equals(Model.MODEL)) {
      throw new Error(
        `Invalid 'model' metadata property in Model stream: ${modelStreamId.toString()}`
      )
    }

    const metadata = { controllers: [controller], model: modelStreamId }
    const state: StreamState = {
      type: Model.STREAM_TYPE_ID,
      content: payload.data,
      metadata,
      signature: SignatureStatus.SIGNED,
      anchorStatus: AnchorStatus.NOT_REQUESTED,
      log: [{ cid: commitData.cid, type: CommitType.GENESIS }],
    }

    await this._schemaValidator.validateSchema(state.content.schema)
    if (state.content.views) {
      this._viewsValidator.validateViews(state.content.views, state.content.schema)
    }

    return state
  }

  /**
   * Applies anchor commit
   * @param commitData - Anchor commit
   * @param state - Document state
   * @private
   */
  async _applyAnchor(commitData: CommitData, state: StreamState): Promise<StreamState> {
    return applyAnchorCommit(commitData, state)
  }
}
