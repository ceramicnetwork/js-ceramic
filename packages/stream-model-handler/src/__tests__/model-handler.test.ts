import { jest } from '@jest/globals'
import { CID } from 'multiformats/cid'
import { decode as decodeMultiHash } from 'multiformats/hashes/digest'
import * as dagCBOR from '@ipld/dag-cbor'
import type { DID } from 'dids'
import { wrapDocument } from '@ceramicnetwork/3id-did-resolver'
import * as KeyDidResolver from 'key-did-resolver'
import { ModelHandler } from '../model-handler.js'
import * as uint8arrays from 'uint8arrays'
import * as sha256 from '@stablelib/sha256'
import cloneDeep from 'lodash.clonedeep'
import { Model, ModelDefinition } from '@ceramicnetwork/stream-model'
import {
  CeramicApi,
  CommitType,
  Context,
  SignedCommitContainer,
  IpfsApi,
  CeramicSigner,
  GenesisCommit,
  RawCommit,
} from '@ceramicnetwork/common'
import { parse as parseDidUrl } from 'did-resolver'

jest.unstable_mockModule('did-jwt', () => {
  return {
    // TODO - We should test for when this function throws as well
    // Mock: Blindly accept a signature
    verifyJWS: (): void => {
      return
    },
    // And these functions are required for the test to run ¯\_(ツ)_/¯
    resolveX25519Encrypters: () => {
      return []
    },
    createJWE: () => {
      return {}
    },
  }
})

const hash = (data: string): CID => {
  const body = uint8arrays.concat([
    uint8arrays.fromString('1220', 'base16'),
    sha256.hash(uint8arrays.fromString(data)),
  ])
  return CID.create(1, 0x12, decodeMultiHash(body))
}

const FAKE_CID_1 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts33jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_2 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts44jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_3 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts55jqcuam7bmye2pb54adnrtccjlsu')
const FAKE_CID_4 = CID.parse('bafybeig6xv5nwphfmvcnektpnojts66jqcuam7bmye2pb54adnrtccjlsu')
const DID_ID = 'did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki'

const jwsForVersion0 = {
  payload: 'bbbb',
  signatures: [
    {
      protected:
        'eyJraWQiOiJkaWQ6MzprMnQ2d3lmc3U0cGcwdDJuNGo4bXMzczMzeHNncWpodHRvMDRtdnE4dzVhMnY1eG80OGlkeXozOGw3eWRraT92ZXJzaW9uPTAjc2lnbmluZyIsImFsZyI6IkVTMjU2SyJ9',
      signature: 'cccc',
    },
  ],
}

const jwsForVersion1 = {
  payload: 'bbbb',
  signatures: [
    {
      protected:
        'ewogICAgImtpZCI6ImRpZDozOmsydDZ3eWZzdTRwZzB0Mm40ajhtczNzMzN4c2dxamh0dG8wNG12cTh3NWEydjV4bzQ4aWR5ejM4bDd5ZGtpP3ZlcnNpb249MSNzaWduaW5nIgp9',
      signature: 'cccc',
    },
  ],
}

const PLACEHOLDER_CONTENT = { name: 'myModel' }

const FINAL_CONTENT: ModelDefinition = {
  name: 'MyModel',
  accountRelation: { type: 'list' },
  schema: {
    $schema: 'https://json-schema.org/draft/2020-12/schema',
    type: 'object',
    properties: {
      stringPropName: {
        type: 'string',
        maxLength: 80,
      },
    },
    additionalProperties: false,
    required: ['stringPropName'],
  },
}

const FINAL_CONTENT_WITH_ACCOUNT_DOCUMENT_VIEW: ModelDefinition = {
  ...FINAL_CONTENT,
  views: {
    owner: { type: 'documentAccount' },
  },
}

const CONTENT_WITH_INVALID_VIEWS: ModelDefinition = {
  ...FINAL_CONTENT,
  views: {
    stringPropName: { type: 'documentAccount' },
  },
}

const CONTENT_WITH_INVALID_SCHEMA = {
  name: 'MyModel',
  accountRelation: { type: 'list' },
  schema: {
    $schema: 'https://json-schema.org/draft/2020-12/schema',
    type: 'object',
    properties: {
      stringPropName: {
        type: 'CLEARLY_A_WRONG_TYPE',
      },
    },
    $defs: ['$DEFS_SHOULD_BE_AN_OBJECT'],
    additionalProperties: false,
    required: 'THIS_SHOULD_BE_AN_ARRAY_OF_STRINGS',
  },
}

const ThreeIdResolver = {
  '3': async (did) => ({
    didResolutionMetadata: { contentType: 'application/did+json' },
    didDocument: wrapDocument(
      {
        publicKeys: {
          signing: 'zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV',
          encryption: 'z6LSfQabSbJzX8WAm1qdQcHCHTzVv8a2u6F7kmzdodfvUCo9',
        },
      },
      did
    ),
    didDocumentMetadata: {},
  }),
}

const setDidToNotRotatedState = (did: DID) => {
  const keyDidResolver = KeyDidResolver.getResolver()
  did.setResolver({
    ...keyDidResolver,
    ...ThreeIdResolver,
  })

  did.createJWS = async () => jwsForVersion0
}

// TODO: De-dupe this with similar code from tile-document-handler.test.ts and model-instance-document.test.ts
const rotateKey = (did: DID, rotateDate: string) => {
  did.resolve = async (didUrl) => {
    const { did } = parseDidUrl(didUrl)
    const isVersion0 = /version=0/.exec(didUrl)

    if (isVersion0) {
      return {
        didResolutionMetadata: { contentType: 'application/did+json' },
        didDocument: wrapDocument(
          {
            publicKeys: {
              signing: 'zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV',
              encryption: 'z6LSfQabSbJzX8WAm1qdQcHCHTzVv8a2u6F7kmzdodfvUCo9',
            },
          },
          did
        ),
        didDocumentMetadata: {
          nextUpdate: rotateDate,
        },
      }
    }

    return {
      didResolutionMetadata: { contentType: 'application/did+json' },
      didDocument: wrapDocument(
        {
          publicKeys: {
            signing: 'zQ3shwsCgFanBax6UiaLu1oGvM7vhuqoW88VBUiUTCeHbTeTV',
            encryption: 'z6MkjKeH8SgVAYCvTBoyxx7uRJFGM2a9HUeFwfJfd6ctuA3X',
          },
        },
        did
      ),
      didDocumentMetadata: {
        updated: rotateDate,
      },
    }
  }

  did.createJWS = async () => jwsForVersion1
}

async function checkSignedCommitMatchesExpectations(
  did: DID,
  commit: SignedCommitContainer,
  expectedCommit: GenesisCommit | RawCommit
) {
  const { jws, linkedBlock } = commit
  expect(jws).toBeDefined()
  expect(linkedBlock).toBeDefined()

  const payload = dagCBOR.decode(linkedBlock)

  const unpacked = { jws, linkedBlock: payload }

  const expected = await did.createDagJWS(expectedCommit)
  expect(expected).toBeDefined()

  const { jws: eJws, linkedBlock: eLinkedBlock } = expected
  const ePayload = dagCBOR.decode(eLinkedBlock)
  const signed = { jws: eJws, linkedBlock: ePayload }

  expect(unpacked).toEqual(signed)
}

describe('ModelHandler', () => {
  let did: DID
  let handler: ModelHandler
  let context: Context
  let signerUsingNewKey: CeramicSigner
  let signerUsingOldKey: CeramicSigner

  beforeAll(async () => {
    process.env.CERAMIC_ENABLE_EXPERIMENTAL_COMPOSE_DB = 'true'

    const recs: Record<string, any> = {}
    const ipfs = {
      dag: {
        put(rec: any, cid?: CID): any {
          if (cid) {
            recs[cid.toString()] = { value: rec }
            return cid
          }
          // stringify as a way of doing deep copy
          const clone = cloneDeep(rec)
          const c = hash(JSON.stringify(clone))
          recs[c.toString()] = { value: clone }
          return c
        },
        get(cid: any): any {
          return recs[cid.toString()]
        },
      },
    } as IpfsApi

    const keyDidResolver = KeyDidResolver.getResolver()
    const { DID } = await import('dids')
    did = new DID({
      resolver: {
        ...keyDidResolver,
      },
    })
    ;(did as any)._id = DID_ID
    const api = {
      getSupportedChains: jest.fn(async () => {
        return ['fakechain:123']
      }),
      did,
    }

    signerUsingNewKey = { did: new DID({}) }
    ;(signerUsingNewKey.did as any)._id = DID_ID
    signerUsingNewKey.did.createJWS = async () => jwsForVersion1

    signerUsingOldKey = { did: new DID({}) }
    ;(signerUsingOldKey.did as any)._id = DID_ID
    signerUsingOldKey.did.createJWS = async () => jwsForVersion0

    context = {
      did,
      ipfs,
      anchorService: null,
      api: api as unknown as CeramicApi,
    }
  })

  beforeEach(() => {
    handler = new ModelHandler()

    setDidToNotRotatedState(did)
  })

  it('is constructed correctly', async () => {
    expect(handler.name).toEqual('model')
  })

  it('makes genesis commits correctly', async () => {
    const commit = await Model._makeGenesis(context.api, FINAL_CONTENT)
    expect(commit).toBeDefined()

    const expectedGenesis = {
      data: FINAL_CONTENT,
      header: { controllers: [context.api.did.id], model: Model.MODEL.bytes },
    }

    await checkSignedCommitMatchesExpectations(did, commit, expectedGenesis)
  })

  it('supports view properties in genesis commit', async () => {
    const commit = await Model._makeGenesis(context.api, FINAL_CONTENT_WITH_ACCOUNT_DOCUMENT_VIEW)
    expect(commit).toBeDefined()

    const expectedGenesis = {
      data: FINAL_CONTENT_WITH_ACCOUNT_DOCUMENT_VIEW,
      header: { controllers: [context.api.did.id], model: Model.MODEL.bytes },
    }

    await checkSignedCommitMatchesExpectations(did, commit, expectedGenesis)
  })

  it('Content is required', async () => {
    await expect(Model._makeGenesis(context.api, null)).rejects.toThrow(
      /Genesis content cannot be null/
    )
  })

  it('creates genesis commits deterministically', async () => {
    const commit1 = await Model._makeGenesis(context.api, FINAL_CONTENT)
    const commit2 = await Model._makeGenesis(context.api, FINAL_CONTENT)

    expect(commit1).toEqual(commit2)
  })

  it('applies genesis commit correctly', async () => {
    const commit = (await Model._makeGenesis(context.api, FINAL_CONTENT)) as SignedCommitContainer
    await context.ipfs.dag.put(commit, FAKE_CID_1)

    const payload = dagCBOR.decode(commit.linkedBlock)
    await context.ipfs.dag.put(payload, commit.jws.link)

    const commitData = {
      cid: FAKE_CID_1,
      type: CommitType.GENESIS,
      commit: payload,
      envelope: commit.jws,
    }
    const streamState = await handler.applyCommit(commitData, context)
    expect(streamState).toMatchSnapshot()
  })

  it('applies genesis commits with views properties correctly', async () => {
    const commit = (await Model._makeGenesis(
      context.api,
      FINAL_CONTENT_WITH_ACCOUNT_DOCUMENT_VIEW
    )) as SignedCommitContainer
    await context.ipfs.dag.put(commit, FAKE_CID_1)

    const payload = dagCBOR.decode(commit.linkedBlock)
    await context.ipfs.dag.put(payload, commit.jws.link)

    const commitData = {
      cid: FAKE_CID_1,
      type: CommitType.GENESIS,
      commit: payload,
      envelope: commit.jws,
    }
    const streamState = await handler.applyCommit(commitData, context)
    expect(streamState).toMatchSnapshot()
  })

  it('fails to apply genesis commits with invalid schema', async () => {
    const commit = (await Model._makeGenesis(
      context.api,
      CONTENT_WITH_INVALID_SCHEMA
    )) as SignedCommitContainer
    await context.ipfs.dag.put(commit, FAKE_CID_1)

    const payload = dagCBOR.decode(commit.linkedBlock)
    await context.ipfs.dag.put(payload, commit.jws.link)

    const commitData = {
      cid: FAKE_CID_1,
      type: CommitType.GENESIS,
      commit: payload,
      envelope: commit.jws,
    }
    await expect(handler.applyCommit(commitData, context)).rejects.toThrow(
      `Validation Error: data/$defs must be object, data/properties/stringPropName/type must be equal to one of the allowed values, data/properties/stringPropName/type must be array, data/properties/stringPropName/type must match a schema in anyOf, data/required must be array`
    )
  })

  it(`fails to apply genesis commits if views validation fails`, async () => {
    const commit = (await Model._makeGenesis(
      context.api,
      CONTENT_WITH_INVALID_VIEWS
    )) as SignedCommitContainer
    await context.ipfs.dag.put(commit, FAKE_CID_1)

    const payload = dagCBOR.decode(commit.linkedBlock)
    await context.ipfs.dag.put(payload, commit.jws.link)

    const commitData = {
      cid: FAKE_CID_1,
      type: CommitType.GENESIS,
      commit: payload,
      envelope: commit.jws,
    }
    await expect(handler.applyCommit(commitData, context)).rejects.toThrow(
      /view definition used with a property also present in schema/
    )
  })

  it('fails to apply signed commit', async () => {
    await expect(handler.applyCommit({}, context, {})).rejects.toThrow(
      `Cannot update a finalized Model`
    )
  })

  it('fails to apply genesis commits with extra fields', async () => {
    const commit = (await Model._makeGenesis(context.api, {
      ...PLACEHOLDER_CONTENT,
      foo: 'bar',
    })) as SignedCommitContainer
    await context.ipfs.dag.put(commit, FAKE_CID_1)

    const payload = dagCBOR.decode(commit.linkedBlock)
    await context.ipfs.dag.put(payload, commit.jws.link)

    const commitData = {
      cid: FAKE_CID_1,
      type: CommitType.GENESIS,
      commit: payload,
      envelope: commit.jws,
    }
    await expect(handler.applyCommit(commitData, context)).rejects.toThrow(
      `Unexpected key 'foo' found in content for Model Stream`
    )
  })

  it('throws error if commit signed by wrong DID', async () => {
    const genesisCommit = (await Model._makeGenesis(context.api, FINAL_CONTENT, {
      controller: 'did:3:fake',
    })) as SignedCommitContainer
    await context.ipfs.dag.put(genesisCommit, FAKE_CID_1)

    const payload = dagCBOR.decode(genesisCommit.linkedBlock)
    await context.ipfs.dag.put(payload, genesisCommit.jws.link)

    const genesisCommitData = {
      cid: FAKE_CID_1,
      type: CommitType.GENESIS,
      commit: payload,
      envelope: genesisCommit.jws,
      timestamp: Date.now(),
    }
    await expect(handler.applyCommit(genesisCommitData, context)).rejects.toThrow(
      /invalid_jws: not a valid verificationMethod for issuer/
    )
  })

  it('applies anchor commit correctly', async () => {
    const genesisCommit = (await Model._makeGenesis(
      context.api,
      FINAL_CONTENT
    )) as SignedCommitContainer
    await context.ipfs.dag.put(genesisCommit, FAKE_CID_1)

    const payload = dagCBOR.decode(genesisCommit.linkedBlock)
    await context.ipfs.dag.put(payload, genesisCommit.jws.link)

    // apply genesis
    const genesisCommitData = {
      cid: FAKE_CID_1,
      type: CommitType.GENESIS,
      commit: payload,
      envelope: genesisCommit.jws,
    }
    let state = await handler.applyCommit(genesisCommitData, context)

    // apply anchor
    const anchorProof = {
      blockNumber: 123456,
      blockTimestamp: 1615799679,
      chainId: 'fakechain:123',
    }
    await context.ipfs.dag.put(anchorProof, FAKE_CID_3)
    const anchorCommitData = {
      cid: FAKE_CID_4,
      type: CommitType.ANCHOR,
      commit: { proof: FAKE_CID_3, id: FAKE_CID_1, prev: FAKE_CID_1 },
      proof: anchorProof,
    }
    state = await handler.applyCommit(anchorCommitData, context, state)
    expect(state).toMatchSnapshot()
  })

  it('fails to apply commit if new key used before rotation', async () => {
    const rotateDate = new Date('2022-03-11T21:28:07.383Z')

    // make genesis with new key
    const genesisCommit = (await Model._makeGenesis(
      signerUsingNewKey,
      FINAL_CONTENT
    )) as SignedCommitContainer
    await context.ipfs.dag.put(genesisCommit, FAKE_CID_1)

    const payload = dagCBOR.decode(genesisCommit.linkedBlock)
    await context.ipfs.dag.put(payload, genesisCommit.jws.link)

    // commit is applied 1 hour before rotation
    const genesisCommitData = {
      cid: FAKE_CID_1,
      type: CommitType.GENESIS,
      commit: payload,
      envelope: genesisCommit.jws,
      timestamp: rotateDate.valueOf() / 1000 - 60 * 60,
    }

    rotateKey(did, rotateDate.toISOString())

    await expect(handler.applyCommit(genesisCommitData, context)).rejects.toThrow(
      /invalid_jws: signature authored before creation of DID version/
    )
  })

  it('applies commit made using an old key if it is applied within the revocation period', async () => {
    const rotateDate = new Date('2022-03-11T21:28:07.383Z')
    rotateKey(did, rotateDate.toISOString())

    // make genesis commit using old key
    const genesisCommit = (await Model._makeGenesis(
      signerUsingOldKey,
      FINAL_CONTENT
    )) as SignedCommitContainer
    await context.ipfs.dag.put(genesisCommit, FAKE_CID_1)

    const payload = dagCBOR.decode(genesisCommit.linkedBlock)
    await context.ipfs.dag.put(payload, genesisCommit.jws.link)

    // commit is applied 1 hour after the rotation
    const genesisCommitData = {
      cid: FAKE_CID_1,
      type: CommitType.GENESIS,
      commit: payload,
      envelope: genesisCommit.jws,
      timestamp: rotateDate.valueOf() / 1000 + 60 * 60,
    }
    const state = await handler.applyCommit(genesisCommitData, context)
    expect(state).toMatchSnapshot()
  })
})
