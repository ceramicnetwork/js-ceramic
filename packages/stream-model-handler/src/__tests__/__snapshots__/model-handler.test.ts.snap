// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModelHandler applies anchor commit correctly 1`] = `
Object {
  "anchorProof": Object {
    "blockNumber": 123456,
    "blockTimestamp": 1615799679,
    "chainId": "fakechain:123",
  },
  "anchorStatus": 3,
  "content": Object {
    "accountRelation": Object {
      "type": "list",
    },
    "name": "MyModel",
    "schema": Object {
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "additionalProperties": false,
      "properties": Object {
        "stringPropName": Object {
          "maxLength": 80,
          "type": "string",
        },
      },
      "required": Array [
        "stringPropName",
      ],
      "type": "object",
    },
  },
  "log": Array [
    Object {
      "cid": Object {
        "code": 112,
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "timestamp": 1615799679,
      "type": 0,
    },
    Object {
      "cid": Object {
        "code": 112,
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          123,
          201,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "timestamp": 1615799679,
      "type": 2,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
    ],
    "model": StreamID {
      "_cid": Object {
        "code": 113,
        "hash": Uint8Array [
          113,
          11,
          0,
          9,
          104,
          109,
          111,
          100,
          101,
          108,
          45,
          118,
          49,
        ],
        "version": 1,
      },
      "_tag": Symbol(@ceramicnetwork/streamid/StreamID),
      "_type": 4,
    },
  },
  "signature": 2,
  "type": 2,
}
`;

exports[`ModelHandler applies commit made using an old key if it is applied within the revocation period 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "accountRelation": Object {
      "type": "list",
    },
    "name": "MyModel",
    "schema": Object {
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "additionalProperties": false,
      "properties": Object {
        "stringPropName": Object {
          "maxLength": 80,
          "type": "string",
        },
      },
      "required": Array [
        "stringPropName",
      ],
      "type": "object",
    },
  },
  "log": Array [
    Object {
      "cid": Object {
        "code": 112,
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
    ],
    "model": StreamID {
      "_cid": Object {
        "code": 113,
        "hash": Uint8Array [
          113,
          11,
          0,
          9,
          104,
          109,
          111,
          100,
          101,
          108,
          45,
          118,
          49,
        ],
        "version": 1,
      },
      "_tag": Symbol(@ceramicnetwork/streamid/StreamID),
      "_type": 4,
    },
  },
  "signature": 2,
  "type": 2,
}
`;

exports[`ModelHandler applies genesis commit correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "accountRelation": Object {
      "type": "list",
    },
    "name": "MyModel",
    "schema": Object {
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "additionalProperties": false,
      "properties": Object {
        "stringPropName": Object {
          "maxLength": 80,
          "type": "string",
        },
      },
      "required": Array [
        "stringPropName",
      ],
      "type": "object",
    },
  },
  "log": Array [
    Object {
      "cid": Object {
        "code": 112,
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
    ],
    "model": StreamID {
      "_cid": Object {
        "code": 113,
        "hash": Uint8Array [
          113,
          11,
          0,
          9,
          104,
          109,
          111,
          100,
          101,
          108,
          45,
          118,
          49,
        ],
        "version": 1,
      },
      "_tag": Symbol(@ceramicnetwork/streamid/StreamID),
      "_type": 4,
    },
  },
  "signature": 2,
  "type": 2,
}
`;

exports[`ModelHandler applies genesis commits with views properties correctly 1`] = `
Object {
  "anchorStatus": 0,
  "content": Object {
    "accountRelation": Object {
      "type": "list",
    },
    "name": "MyModel",
    "schema": Object {
      "$schema": "https://json-schema.org/draft/2020-12/schema",
      "additionalProperties": false,
      "properties": Object {
        "stringPropName": Object {
          "maxLength": 80,
          "type": "string",
        },
      },
      "required": Array [
        "stringPropName",
      ],
      "type": "object",
    },
    "views": Object {
      "owner": Object {
        "type": "documentAccount",
      },
    },
  },
  "log": Array [
    Object {
      "cid": Object {
        "code": 112,
        "hash": Uint8Array [
          18,
          32,
          222,
          189,
          122,
          219,
          60,
          229,
          101,
          68,
          210,
          42,
          111,
          107,
          147,
          57,
          111,
          105,
          128,
          168,
          6,
          124,
          44,
          193,
          52,
          240,
          247,
          128,
          27,
          99,
          49,
          9,
          43,
          149,
        ],
        "version": 1,
      },
      "type": 0,
    },
  ],
  "metadata": Object {
    "controllers": Array [
      "did:3:k2t6wyfsu4pg0t2n4j8ms3s33xsgqjhtto04mvq8w5a2v5xo48idyz38l7ydki",
    ],
    "model": StreamID {
      "_cid": Object {
        "code": 113,
        "hash": Uint8Array [
          113,
          11,
          0,
          9,
          104,
          109,
          111,
          100,
          101,
          108,
          45,
          118,
          49,
        ],
        "version": 1,
      },
      "_tag": Symbol(@ceramicnetwork/streamid/StreamID),
      "_type": 4,
    },
  },
  "signature": 2,
  "type": 2,
}
`;
